api/src/main/java/marquez/db/LineageDao.java

```diff
@@ -154,4 +171,41 @@ SELECT DISTINCT on(r.job_name, r.namespace_name) r.*, jv.version as job_version
       WHERE j.uuid in (<jobUuid>) OR j.symlink_target_uuid IN (<jobUuid>)
       ORDER BY r.job_name, r.namespace_name, created_at DESC""")
   List<Run> getCurrentRuns(@BindList Collection<UUID> jobUuid);
+
+  @SqlQuery(
+      """
+WITH RECURSIVE
+  upstream_runs(
+          r_uuid, started_at, ended_at, state,
+          job_uuid, job_version_uuid, job_namespace, job_name,
+          dataset_uuid, dataset_version_uuid, dataset_namespace, dataset_name,
+          u_r_uuid, depth) AS (
+    SELECT
+          r.uuid, r.started_at, r.ended_at, r.current_run_state,
+          r.job_uuid, r.job_version_uuid, r.namespace_name, r.job_name,
+          dv.dataset_uuid, dv."version", dv.namespace_name, dv.dataset_name,
+          dv.run_uuid,
+          0 AS depth
+    FROM runs r
+    LEFT JOIN runs_input_mapping rim ON rim.run_uuid = r.uuid
+    LEFT JOIN dataset_versions dv ON dv.uuid = rim.dataset_version_uuid
+    LEFT JOIN runs r1 ON r1.uuid = dv.run_uuid
+    WHERE r.uuid = :runId
+  UNION
+    SELECT
+          ur.u_r_uuid, r2.started_at, r2.ended_at, r2.current_run_state,
+          r2.job_uuid, r2.job_version_uuid, r2.namespace_name, r2.job_name,
+          dv2.dataset_uuid, dv2."version", dv2.namespace_name, dv2.dataset_name,
+          dv2.run_uuid,
+          ur.depth + 1 AS depth
+    FROM upstream_runs ur
+    INNER JOIN runs r2 ON r2.uuid = ur.u_r_uuid
```

Comment:
`runs` table can get huge. I think it doable to construct here recursion steps on the `runs_input_mapping` table only.

Reply:
This is where I find the job information (r2.job_uuid, r2.job_version_uuid, r2.namespace_name, r2.job_name). Do you think I could get it a different way?
unless we add it to runs_input_mapping?

Reply:
But I think I could actually move the runs join out of the recursion into the final select statement.

Reply:
I just updated the query to move the joins on the `runs` table outside of the recursion and in the final select statement.